sys/kern/sys_process.c

```diff
@@ -643,7 +643,7 @@ sys_ptrace(struct thread *td, struct ptrace_args *uap)
 	 * structs may be too large to put on the stack anyway.
 	 */
 	union {
-		struct ptrace_io_desc piod;
+		struct ptrace_io_desc_c piod;
```

Comment:
You could fix `piod_offs` and `piod_addr` to be `__kerncap` in `struct ptrace_io_desc` instead of using `struct ptrace_io_desc_c`

Reply:
So, now I should add ifdefs with COMPAT_FREEBSD64 in kern_ptrace and assume that the default one is a purecap userspace, right?

Reply:
and also assume that `__has_feature(capabilities)` is always true?

Reply:
COMPAT_FREEBSD64, yes, if it's needed for any of the cases where COMPAT_FREEBSD32 additions already exist. But you can't assume we have capabilities; CheriBSD supports being compiled without CHERI available (both not enabling CHERI on MIPS/RISC-V, and building on architectures where we don't support CHERI at all).

Reply:
nice, thanks for the clarification. Then quite a lot of changes are needed...

Reply:
I tried in the new commit. Fasttrap works with the changes, but it is not a comprehensive test.